homeassistant/helpers/event.py

```diff
@@ -103,12 +100,91 @@ def state_change_listener(event: Event) -> None:
                 event.data.get("new_state"),
             )
 
+    if entity_ids != MATCH_ALL:
+        # If we have a list of entity ids we use
+        # async_track_state_change_event to route
+        # by entity_id to avoid iterating though state change
+        # events and creating a jobs where the most
+        # common outcome is to return right away because
+        # the entity_id does not match since usually
+        # only one or two listeners want that specific
+        # entity_id.
+        return async_track_state_change_event(hass, entity_ids, state_change_listener)
+
     return hass.bus.async_listen(EVENT_STATE_CHANGED, state_change_listener)
 
 
 track_state_change = threaded_listener_factory(async_track_state_change)
 
 
+@bind_hass
+def async_track_state_change_event(
+    hass: HomeAssistant, entity_ids: Iterable[str], action: Callable[[Event], None]
+) -> Callable[[], None]:
+    """Track specific state change events indexed by entity_id.
+
+    Unlike async_track_state_change, async_track_state_change_event
+    passes the full event to the callback.
+
+    In order to avoid having to iterate a long list
+    of EVENT_STATE_CHANGED and fire and create a job
+    for each one, we keep a dict of entity ids that
+    care about the state change events so we can
+    do a fast dict lookup to route events.
+    """
+
+    entity_callbacks = hass.data.setdefault(TRACK_STATE_CHANGE_CALLBACKS, {})
+
+    if TRACK_STATE_CHANGE_LISTENER not in hass.data:
+
+        @callback
+        def _async_state_change_dispatcher(event: Event) -> None:
+            """Dispatch state changes by entity_id."""
+            entity_id = event.data.get("entity_id")
+
+            if entity_id not in entity_callbacks:
+                return
+
+            for action in entity_callbacks[entity_id]:
+                hass.async_run_job(action, event)
```

Comment:
Of course I realize something the moment I press the merge button: `async_run_job` will directly call functions decorated with `@callback`. Now if one of those functions raises, it will cancel any subsequent entity changed callbacks.

Reply:
https://github.com/home-assistant/core/pull/37179